### ISSUE TITLE
Current chunking strategy may lead to serious hallucination.

### ISSUE THREAD

**Issue by kkaiwwana** [AMAAI-Lab/SonicVerse#2]
Hi there, sorry for opening another issue today. As #1  is solved, I just start playing with SonicVerse. It's pretty interesting. However, here is another issue resulted by current chunking strategy ([code](https://huggingface.co/spaces/amaai-lab/SonicVerse/blob/main/app.py#L105-L128)).  

**TL;DR** For those music of size slightly longer than N * chunk_size, the last tiny chunk (mostly empty) may trigger LLM's hallucination and therefore produce false caption.

Full info is provided below. Recently I'm playing with musictengo too. So I'm just cruious whether the caption generated by SonicVerse of mustango generated music would be consisent with its prompt. I start with an official [exmaple](https://amaai-lab.github.io/mustango/samples/samples2/output_71.wav). It's length is actually a little bit longger than 10s (technically, 10.242s or 163872 samples over 16000 sample rate). So the function `split_audio` splits it into 2 chunks (10s + 0.242s), and here is the generated caption:

![Image](https://github.com/user-attachments/assets/f32c9557-e114-40e2-90a3-50c3bed1652f)

The description of `vocalist` in the third and forth paragraph is clearly not true. And this issue persists across multiple runs (same `vocalist`, not gonna attach here for clarity). I realized it's the problem with chunking and check the code related. Then I just trimmed it to 10s (something like 9.99), and it did produce "expected" caption as follows:

![Image](https://github.com/user-attachments/assets/8ba5f237-9042-494a-9508-8ce75d7c4367)

So, I suppose a more robust chunking strategy may be needed, e.g. make N-th piece overlapped with (N-1)-th, or just drop the short tail.

**Comment by kkaiwwana** [comment]
And here is another minor issue: it seems that the caption of mustango generated music was not very consisent with prompt e.g. chords, key.

In example mentioned above, its prompt is:  "*This techno song features a synth lead playing the main melody. This is accompanied by programmed percussion playing a simple kick focused beat. The hi-hat is accented in an open position on the 3-and count of every bar. The synth plays the bass part with a voicing that sounds like a cello. This techno song can be played in a club. The chord sequence is Gm, A7, Eb, Bb, C, F, Gm. The beat counts to 2. The tempo of this song is 128.0 beats per minute. The key of this song is G minor.*"

Another shot of caption is like:

![Image](https://github.com/user-attachments/assets/4e7cdf07-783d-469a-9cb7-5c5d1e8cb4c0)

I understand how hard it's (for both generation and caption). I just curious about if it's the problem with generated music or it's actually the problem in caption model side. (sadly I struggle to identify chords and key by ears)

**Comment by dorienh** [comment]
Hi, thanks for all your interest in our work. 

@annabeth97c Maybe a small check to see if the chunk is at least 4s or so would be wise? Otherwise discart?

In the SonicVerse paper, we actually evaluate the model on MusicBench (which was used to train Mustango) and it performed reasonably. However, it was not trained on MusicBench, hence it learned another 'style' of caption.

**Comment by kkaiwwana** [comment]
Thanks for the reply. But wasn't it trained on MusicBench really? I look through the paper and it says "*We use three datasets for training the proposed architecture...Finally, we use the MusicBench training set...*" in Sec 4.1; And the other two training datasets do not include chord progression, so I suppose the model is actually trained on MusicBench and therefore learns to generate those chord stuffs (not evaluated though). Is that right? Or did I just misunderstand it?

And here is another concern about the chunking problem: would it hurt the evaluation results by any chance? As SonicVerse was evaluated on MusciBench test set, and it contains mostly 10s music, right? So, are those music also a little bit longer than 10s? I suppose that may hinder the model's performance just as above demonstrated. I didn't have MusicBench downloaded on my current device right now. I could perhaps check it when I get home tonight, or would someone on your side be willing to take a look? (If this concern is indeed valid.)

**Comment by dorienh** [comment]
Sorry I should have looked up the paper again. I mostly meant to say that the 10s model was _evaluated_ on MusicBench eval, but indeed the _training_ set of MusicBench was also used for training the model. 

Both our subjective/objective evaluation in the paper focuses on the generative 10s model that we trained (without the chaining). In future work we will dive deeper into the chaining part, which at the moment relies on expert evaluation only if I am not mistaken. We hope to solidify the results in the future (and we are open to collaborators btw!). 

As for **chords**, we also augmented the other datasets _"Because the datasets lack ground-truth music feature labels, we augment them using MIRFLEX, which generated labels for key, instrument, mood/theme, genre, vocals and vocals gender, chords as well as downbeat timings"_ This is used in our multitask setup, not to augment the captions. So the model basically learns to predict chords, and can then decide to use it in the caption or not. The reason we did not include Chords in the evaluation table is because it is very hard to capture (and check similarity) between long chord sequences. When creating MIDICaps, we spent a long time thinking of how to best pull out 'the most prominent chord sequence', but it can be subjective and hard to evaluate. So we opted to leave it out. Since MusicBench eval has chords, they are implicitly evaluated in Table 1.

@annabeth97c if you have time today, can you look into this, or at least set a min. length for chunks?

**Comment by kkaiwwana** [comment]
Yes, even though things like emotion and genre in music seem to be well reflected in the caption already, I can still see the challenge in accurately handling those structured stuffs like chord/beat, particularly under current end2end setup. And I actually do hope to contribute to AI for music in the future. TBH, though I probably prefer not to mention this in a public thread, I've been giving a lot of thought lately to making AI for music the focus of my future phd and that's why I started paying more attention to AI for music projects recently.

**Comment by dorienh** [comment]
Yes, it is a challenge indeed! If you ever want to collaborate, do feel free to send me a PM or email to discuss.
